### PR TITLE
Add fund subcommand

### PR DESCRIPTION
**Pull request by ruyadorno**
## Overview

This PR introduces the `npm fund` command that lists all `funding` info provided by the installed dependencies of a given project.

## :pencil2: Changes
- `lib/utils/funding.js` Provides helpers to validate funding info and
return a tree-shaped structure containing the funding data for all deps.
- `lib/fund.js` Implements `npm fund <pkg>` command
- Added tests
  - `npm install` mention of funding
  - `npm fund <pkg>` variations
  - unit tests for added `lib/utils` and `lib/install` helpers
- Added docs for `npm fund`, `funding` `package.json` property
- Fixed `lib/utils/open-url` to support `--json` config
- Documented `unicode` on `npm install` docs

## :link: References
- RFC: https://github.com/npm/rfcs/blob/2d2f00457ab19b3003eb6ac5ab3d250259fd5a81/accepted/0017-add-funding-support.md

## :camera_flash: Screenshots

> Simple npm install within a project with given dependencies

Demo showing simple npm install within a project with given dependencies

> npm install a package that contains dependencies with funding declared, fires up `npm ls` and `npm fund` to show differences in output

Demo showing npm install a package that contains dependencies with funding declared, fires up npm ls and npm fund to show differences

> npm fund `--json` output example

Demo showing npm fund --json example

## :mag: Testing

Install a package that has a valid `funding` object declared on its `package.json`, e.g:

- Single package: `npm install sleepover`
- Many nested packages with funding info: `npm install @ruyadorno/test-funding-dep`

✅ This change has unit test coverage
✅ This change has integration test coverage

## :fire: Rollback

> If we observe something wrong with this change in production, how should we respond?

This can be reverted at any time

### CODE REVIEW

**Comment by brody2consult** [comment]
I would personally favor some way to advertise which project maintainers are available for hire with some quick details or have any other services to offer (no third-party ads please).

For example (true example): "@brodybits is available for short-term hire in Boston, NYC, or remote contract"

**Comment by darcyclarke** [comment]
@brodybits Hey Chris! Appreciate the feedback but I think it's out of the scope of this work which was talked about & simplified pretty extensively in the original [RFC](https://github.com/npm/rfcs/blob/2d2f00457ab19b3003eb6ac5ab3d250259fd5a81/accepted/0017-add-funding-support.md). If you have an idea about a future feature/design we could support, definitely feel free to contribute an RFC over on our [npm/rfcs](https://github.com/npm/rfcs/) repo and/or, where I think your suggestion may make the most sense, is to investigate the work the [Node Foundation Package Maintenance Working Group](https://github.com/nodejs/package-maintenance/) has done around a more verbose [`support` schema](https://github.com/nodejs/package-maintenance/blob/master/docs/drafts/PACKAGE-SUPPORT.md).

**Comment by giuseppeg** [comment]
This looks amazing! It would be nice to have the info deduplicated and flattened and maybe have an advanced mode (maybe a `--tree` flag) to show how the packages are related. 

There seem to be a lot of "noise" in the output from the first gif, a flat list with a white line to break every package info would be more readable.

**Comment by ljharb** [comment]
If i haven’t chosen to add a dep to package.json, it’s critical that i know how and why a dep is in my graph, or else why would i ever want to give it money?

**Comment by giuseppeg** [comment]
@ljharb fair but imagine having 50 packages with sub deps. The list could still be flat and each field can say "Used in: babel-core" or something (assuming babel-core is the dependency you installed).

**Comment by ljharb** [comment]
it could, but presumably you’d want the list sorted, with direct deps at the top, and then transitive deps sorted by “used in the most packages”. (iow, I’m partially agreeing with you; a tree isn’t the best way to represent this since duplicated deps should be consolidated and summarized)

**Comment by giuseppeg** [comment]
Lock files (yarn.lock and package-lock.json) are flat, I'd personally go for that. That said probably shipping any solution and asking the community for feedback is the best way to go.

**Comment by coreyfarrell** [comment]
A potential issue is that this is version based rather than package based?  This means that I cannot set funding information for already released versions of a package.  More importantly if funding information changes after a release I cannot fix it (remove a no longer valid funding method for example).

**Comment by ruyadorno** [comment]
@npm/cli-team: @giuseppeg raised an important point that went a bit under documented in my original description which is that the current implementation of `npm fund` output does some smart things when building that tree, such as moving transitive deps up in the tree if their parents has no `funding` info (since it only displays packages that contain such info) along with the stacking of same-maintainer info demoed in the 2nd gif there.

I quickly put together a demo showing how childs will show up a level above compared to a `npm ls` output:

> in this demo the  dependency`@ruyadorno/dep-sub-no-funding@1.0.0` which has no funding info is removed from the `npm fund` listing and its child are listed as children of its parent
Demo showing diff between npm ls and npm fund when a parent dep has no funding info

the takeaway here is that this output is neither an exact representation of the dependency tree nor a flat list, it's something in between and my idea when working on it is that it's a middle-ground to start with and we can iterate later on with feedback from the community

**Comment by brody2consult** [comment]
> it’s critical that i know how and why a dep is in my graph

@ljharb `yarn why` would do that if you use Yarn.

I think this should be supported by npm, looks like there is already a `npm-why` package to do this.

**Comment by ljharb** [comment]
@brodybits i don't need to know that information in general, and i know how to get it without using yarn - i'm talking about *directly inline in the funding info*, which is where i'd need to see it to gain motivation to donate money to a transitive dep.

**Comment by ljharb** [comment]
@ruyadorno and what if `test-funding-dep-sub-sub-sub` is indirectly used by 23 different top-level packages? must i see it listed 23 times, or must its provenance only be shown for one ancestor and the other 22 lost?

**Comment by ruyadorno** [comment]
> or must its provenance only be shown for one ancestor and the other 22 lost?

☝️ that is the case yes, in its current implementation

**Comment by ljharb** [comment]
That seems like proof that this is not the optimal output format to use.

File: lib/utils/funding.js [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+
+  function getFundingDependencies (tree) {
+    const deps = tree && tree.dependencies
+    if (!deps) return empty()
+
+    // broken into two steps to make sure items appearance
+    // within top levels takes precedence over nested ones
+    return (Object.keys(deps)).map((key) => {
+      const dep = deps[key]
+      const { name, funding, version } = dep
```
Comment (znarf) [inline lib/utils/funding.js:73]:
```suggestion
      const { name, version } = dep
      const funding = dep.funding || dep.collective;
```

There are many packages using the `collective` property today to expose their Open Collective URL. Would be great to support that when `funding` is not found, while we wait for them to add it. I

The syntax of `collective` is exactly the same as `funding`, making the implementation uncomplicated.

It's also nice that this allow anyone to test the feature and see results today, while in the current state it's giving me only empty results for all the projects I tested.
